distage/distage-core-api/src/main/scala/izumi/distage/model/definition/dsl/AbstractBindingDefDSL.scala

```diff
@@ -211,8 +211,8 @@ object AbstractBindingDefDSL {
             case ImplDef.ProviderImpl(implType, function) =>
               val oldMagnet = ProviderMagnet(function)
               val newProvider = providerMagnetModifier(oldMagnet).get
-              if (implType == newProvider.ret) {
-                b = b.withImplDef(ImplDef.ProviderImpl(newProvider.ret, newProvider))
+              if (newProvider.ret <:< implType) {
```

Comment:
Are you sure the condition is necessary at all, since we already restrict types? If so we should at least have an `else` branch throwing an exception, so that no silent ignoring happens if the condition doesn't apply.